src/cow-react/common/hooks/useSafeMemo.test.ts

```diff
@@ -0,0 +1,82 @@
+import { useSafeMemo } from './useSafeMemo'
+import { renderHook } from '@testing-library/react-hooks'
+import { useMemo, useState } from 'react'
+import { CurrencyAmount, Token } from '@uniswap/sdk-core'
+import { SupportedChainId as ChainId } from '../../../custom/constants/chains'
+
+function createInputCurrency() {
+  return new Token(ChainId.GOERLI, '0xe583769738b6dd4e7caf8451050d1948be717679', 6, 'USDT', 'Tether USD')
+}
+
+function createOutputCurrency() {
+  return new Token(ChainId.GOERLI, '0xca063a2ab07491ee991dcecb456d1265f842b568', 8, 'WBTC', 'Wrapped BTC')
+}
+
+function createInputAmount() {
+  return CurrencyAmount.fromRawAmount(createInputCurrency(), 100_000)
+}
+
+function createOutputAmount() {
+  return CurrencyAmount.fromRawAmount(createOutputCurrency(), 200_000)
+}
+
+function createStaticObject() {
+  return {
+    inputCurrency: createInputCurrency(),
+    outputCurrency: createOutputCurrency(),
+    inputAmount: createInputAmount(),
+    outputAmount: createOutputAmount(),
+    address: 'xxx',
+  }
+}
+
+describe('useSafeMemo() to avoid redundant actuation of hooks', () => {
+  it('Should execute memo only once using useSafeMemo', () => {
+    let memoCalls = 0
+    let updatesCount = 3
+
+    const { result } = renderHook(() => {
+      const [state, setState] = useState(createStaticObject())
+
+      const memoized = useSafeMemo(Object.values(state), () => {
+        memoCalls++
+        return state
+      })
+
+      if (updatesCount !== 0) {
+        updatesCount--
```

Comment:
I didn't get this flow control. Can you explain?

Reply:
Yes, sure.
1. On 63 line we create a `state` variable
2. On 66 line we create a `memoized` variable, it's the same as `state` but memoized
3. On 73 line we update `state` with the object that contains the same values, but it's not the same object (when compared by reference, it will return false). And we do it 3 times
4. Eventually we check how many times the `useMemo` callback was called

Reply:
I mean, I didn't get why it would be run 3 times.
Now I think I got it. Everytime you call `createStaticObject` it'll update the state to an identical object, but since it's an object it has a different reference and react thinks it is a different object.

Reply:
Yes, exactly :)